test/c/test_sparse_observable.c

```diff
@@ -0,0 +1,650 @@
+
+        for (uint32_t n = 0; n < nni; n++) {
+            // this loop is only called once, so we can use ``n`` to index here
+            bits[n] = view.bit_terms[n];
+            indices[n] = view.indices[n];
+        }
+    }
+
+    qk_obs_free(obs);
+
+    int result = Ok;
+    int expected_nnis[2] = {0, 3};
+    int expected_bits[3] = {QkBitTerm_X, QkBitTerm_Y, QkBitTerm_Z};
+    int expected_indices[3] = {0, 1, 2};
+
+    // check number of terms
+    if (num_terms != 2) {
+        result = EqualityError;
+    }
+
+    // check NNIs
+    for (int i = 0; i < 2; i++) {
+        if (nnis[i] != expected_nnis[i]) {
+            result = EqualityError;
+        }
+    }
+
+    // check bit terms and indices
+    for (int n = 0; n < 3; n++) {
+        if (indices[n] != expected_indices[n] || bits[n] != expected_bits[n]) {
+            result = EqualityError;
+        }
+    }
+
+    return result;
+}
+
+/**
+ * Test copying and modifying a term.
+ */
+int test_copy_term() {
+    // create an observable with the term X0 Y1 Z2
+    u_int32_t num_qubits = 100;
+    QkSparseObservable *obs = qk_obs_zero(num_qubits);
+
+    complex double coeff = 1;
+    QkBitTerm bits[3] = {QkBitTerm_X, QkBitTerm_Y, QkBitTerm_Z};
+    uint32_t indices[3] = {0, 1, 2};
+
+    QkSparseTerm term = {coeff, 3, bits, indices, num_qubits};
+    qk_obs_add_term(obs, &term);
+
+    // now we add a modified copy of the first term -- we use use qk_obs_term(..., &borrowed) on
+    // purpose here
+    QkSparseTerm borrowed;
+    int error = qk_obs_term(obs, 0, &borrowed); // get view on 0th term
+    if (error > 0) {
+        return RuntimeError;
+    }
+
+    // copy the term so we can safely modify it and add it onto the observable
+    size_t len = borrowed.len;
+    QkBitTerm *copied_bits = (QkBitTerm *)malloc(len * sizeof(QkBitTerm));
+    uint32_t *copied_indices = (uint32_t *)malloc(len * sizeof(uint32_t));
+    for (size_t i = 0; i < len; i++) {
+        copied_bits[i] = borrowed.bit_terms[i];
+        copied_indices[i] = borrowed.indices[i];
+    }
+
+    // modify the term and add it onto the observable
+    complex double coeff2 = 2 * I;
+    copied_indices[1] = 99;
+    copied_bits[0] = QkBitTerm_Zero;
+    QkSparseTerm copied = {
+        coeff2, borrowed.len, copied_bits, copied_indices, borrowed.num_qubits,
+    };
+    qk_obs_add_term(obs, &copied);
+
+    free(copied_indices);
+    free(copied_bits);
+
+    // now we directly construct the expected observable
+    QkBitTerm bits2[3] = {QkBitTerm_Zero, QkBitTerm_Y, QkBitTerm_Z};
+    uint32_t indices2[3] = {0, 99, 2};
+    QkSparseTerm term2 = {coeff2, 3, bits2, indices2, num_qubits};
+
+    QkSparseObservable *expected = qk_obs_zero(num_qubits);
+    qk_obs_add_term(expected, &term);
+    qk_obs_add_term(expected, &term2);
+
+    bool equal = qk_obs_equal(expected, obs);
+    qk_obs_free(obs);
+    qk_obs_free(expected);
+
+    if (!equal) {
+        return EqualityError;
+    }
+    return Ok;
```

Comment:
Super minor, but if you like: if you're doing this a lot, C has a ternary conditional operator that works similar to Python's `a if x else b`.  It looks like `x ? a : b`, so you can have `return equal ? Ok : EqualityError;`.

No need to change.

Reply:
This cleaned the code up a bit in 624c3dd 🙂